### ISSUE TITLE
Replace RocResult with roc_panicked

### ISSUE THREAD

**Issue by rtfeldman** [roc-lang/roc#1403]
## Background

Currently, when the host calls a compiled Roc application function, the return value gets wrapped in a `RocResult` which the host needs to unwrap. This works essentially like a `Result` in Rust, and it accounts for the possibility that the Roc application threw an exception (e.g. due to trying to add an element to a `List` that was already at the maximum length).

Now that we require hosts to define `roc_alloc`, `roc_dealloc`, and `roc_realloc`, we can replace the `RocResult` wrapper with the requirement that hosts additionally define a `roc_panicked` function which takes two arguments: an error code `u32` and a `RocStr`. Instead of the compiled application returning `RocResult::Err` if there's an exception, instead it would call `roc_panicked` passing the exception code (a nonzero `u32`) and a `RocStr` with more information when appropriate. 

> **Note:** The original app function call would return memory garbage; it wouldn't bother to set return registers or write into `alloca` pointers, so the host should make sure not to do anything with the return value if `roc_panicked` was called!

## Benefits

`roc_panicked` has a few advantages over the `RocResult` design.

The main advantage is that hosts are no longer required to add an extra conditional every time they call a Roc application function. Since it should be extremely rare in practice that `roc_panicked` is ever invoked, hosts can implement it using something like a C++ exception (such as an unwind or a jump, or an actual C++ `throw`), or even end the program immediately (e.g. `panic!` in Rust). This makes it possible for hosts to call Roc application functions repeatedly in a hot path without paying for an extra conditional every time.

A secondary advantage is that it means the calling convention for calling Roc functions is the same as the C calling convention. Currently, if a Roc host returns an `i64`, it can't return that directly in a register because it has to actually return `RocResult`. Using `roc_panicked` would therefore enable more efficient calls to some compiled application functions. As a bonus, not having to remember to use the `RocResult` wrapper means there's one less thing for host authors to remember.

## Using per-call conditionals with `roc_panicked`

Of note, the `roc_panicked` design still permits having the host do a conditional to check for errors after every call to a Roc application function, just like today. The way to do this would be to have `roc_panicked` set a global variable (or use thread-local storage in a multithreaded context) for something named (for example) `roc_panic_code : u32`, and then another for `roc_panic_str : RocStr`. Initialize `roc_panic_code` to 0 and `roc_panic_str` to an empty `RocStr`.

After making the application call, check `if roc_panic_code == 0` - if it's been changed from its default of 0, that means `roc_panicked` must have been called during the course of the application code running. Now that we know this, we can handle the error appropriately by using a combination of `roc_panic_code` and `roc_panic_str` - just like we previously would have with `RocResult`. If we're able to gracefully recover, we then reset `roc_panic_code` to 0 before the next application call. That's it!

## Summary

In summary, this design makes it possible to have the same "run a conditional after every application call" code we have today (with the same conditional performance in the happy path), but it also permits having faster calls between host and application (due to calling convention improvements) while also opening the door to faster exception-handling mechanisms than running a conditional after every call.
